File: Binary_Tree_Maximum_Path_Sum.java

/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */
class Solution {
    private int ans;

    private int pathSum(TreeNode root) {
        if (root == null)
            return 0;
        int right = pathSum(root.right);
        int left = pathSum(root.left);
        // System.out.printf("%d %d %d\n", right, left, root.val);
        ans = Integer.max(ans, right + left + root.val);
        return Integer.max(Integer.max(right, left) + root.val, 0);
    }

    public int maxPathSum(TreeNode root) {
        if (root == null)
            return 0;
        ans = root.val;
        pathSum(root);
        return ans;
    }
}

public class Binary_Tree_Maximum_Path_Sum {
    public static void main(String[] args) {
        Solution sol = new Solution();
        String[][] testset = { { "1", "2", "3" }, { "-10", "9", "20", "null", "null", "15", "7" }, { "-3" },
                { "1", "-2", "3" } };
        for (String[] test : testset) {
            TreeNode root = TreeUtils.buildTree(test);
            TreeUtils.printTree(root);
            System.out.println(sol.maxPathSum(root));
        }
    }
}

File: Binary_Tree_Maximum_Path_Sum.cpp
#include <iostream>
using namespace std;
// https://leetcode.com/problems/binary-tree-maximum-path-sum/discuss/39775/Accepted-short-solution-in-Java
// Given a non-empty binary tree, find the maximum path sum.
/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
    int maxSum;
    // pathDown
    //  (1) computes the maximum path sum with highest node is the input node, 
    // update maximum if necessary (2) returns the maximum sum of the path that 
    // can be extended to input node's parent.
    int pathDown(TreeNode *node) {
        if (node == NULL)
            return 0;
        int ri = max(0, pathDown(node->right));
        int le = max(0, pathDown(node->left));
        maxSum = max(maxSum, ri + le + node->val);
        // return the sum of a path
        return max(ri, le) + node->val;
    }

  public:
    int maxPathSum(TreeNode *root) {
        if (root == NULL)
            return 0;
        else
            maxSum = root->val;
        pathDown(root);
        return maxSum;
    }
};

// Input: [-10,9,20,null,null,15,7]
//    -10
//    / \
//   9  20
//     /  \
//    15   7
// Output: 42
